File: Heaps/Heaps.java
import java.util.Scanner;


class Heaps {
    public static void main(String[] argv) {
        Scanner sc = new Scanner(System.in);
        int nsites = sc.nextInt();
        int nheaps = sc.nextInt();
        int[] distance = new int[nsites-1];
        int[] weight = new int[nsites];
        int previous_position = sc.nextInt();
        weight[0] = sc.nextInt();
        for(int site=1; site<nsites; site++) {
            int position = sc.nextInt();
            distance[site-1] = position - previous_position;
            previous_position = position;
            weight[site] = sc.nextInt();
        }
        System.out.println(new Problem(distance, weight).solve(nheaps));
    }

    static class Problem {
        // Distance from this site to the next
        int[] distance;
        // Weight of this site's heap
        int[] weight;
        // Number of sites
        int sites;
        // Cost of sub-problems
        // First index: number of sites to visit. From 1 to sites.
        // Second index: number of heaps to assemble. From 1 to sites_left
        // 1-based indexing for prettier code
        int[][] subproblem;

        public Problem(int[] _distance, int[] _weight) {
            distance = _distance;
            weight = _weight;
            sites = weight.length;
        }

        int[][] calcFoldCost(int nheaps) {
            // Calculate cost of folding j heaps starting from site i
            // Maximum fold length is sites-nheaps
            int[][] ret = new int[sites-1][];
            for(int ii=0; ii<sites-1; ii++) {
                int maxfold = Math.min(sites-ii-1, sites-nheaps);
                ret[ii] = new int[maxfold+1];
                ret[ii][0] = 0;
                int barge_load = weight[ii];
                for(int jj=1; jj<ret[ii].length; jj++) {
                    ret[ii][jj] = ret[ii][jj-1] + barge_load * distance[ii+jj-1];
                    barge_load += weight[ii+jj];
                }
            }
            return ret;
        }

        public int solve(int nheaps) {
            int[][] foldCost = calcFoldCost(nheaps);
            subproblem = new int[sites+1][];
            subproblem[0] = new int[] {0};
            // Solve the subproblem for 1..sites sites left to visit
            for(int sites_left=1; sites_left<=sites; sites_left++) {
                int max_heaps = Math.min(nheaps, sites_left);
                int current_site = sites-sites_left;
                subproblem[sites_left] = new int[max_heaps + 1];
                // No need to carry loads if I already have the required number
                // of heaps
                if(max_heaps == sites_left) {
                    subproblem[sites_left][sites_left] = 0;
                    max_heaps--;
                }
                // Solve the subproblem for 1..max_heaps heaps desired
                for(int heaps = Math.max(1,nheaps-current_site); heaps <= max_heaps; heaps++) {
                    // Select the minimum cost out of:
                    // - skipping this site (only an option if heaps != 1)
                    // - taking this site's heap to the next site
                    // - taking this site and the next side's heaps to the
                    // following one
                    // - ...
                    // - taking all the heaps from the next (sites_left-heaps)
                    // sites to the next.
                    //
                    // The cost of each option is the cost of the barge
                    // operation plus the cost of doing the rest of the heaps
                    // (which is a subproblem I've already solved)
                    if(heaps == 1) {
                        subproblem[sites_left][heaps] = foldCost[current_site][sites_left-1];
                        continue;
                    }
                    // If heaps > 1 there's always the option of skipping this
                    // site
                    int mincost = subproblem[sites_left-1][heaps-1];
                    // Calculate the cost of 1..(sites_left-heaps) loads
                    for(int loads=1; loads <= sites_left - heaps; loads++) {
                        int newcost = foldCost[current_site][loads]
                                    + subproblem[sites_left-loads-1][heaps-1];
                        mincost = Math.min(mincost, newcost);
                    }
                    subproblem[sites_left][heaps] = mincost;
                }
            }
            return subproblem[sites][nheaps];
        }
    }

}

File: Heaps/Heaps.cpp
#define NDEBUG 
#include <climits>
#include <cassert>
#include <algorithm>
#include <cstdio>
#include <fstream>

#include "Heaps.h"

Heaps::Heaps(unsigned int sites, cost weight[], dist distance[]) {
    this->sites = sites;
    this->weight = weight;
    this->distance = distance;
}

void Heaps::calcFolds(unsigned int heaps) {
    fold = new cost*[sites];
    for(unsigned int initial=0; initial < sites; initial++) {
        unsigned int foldmax = std::min(sites-heaps+initial, sites-initial);
        // foldmax <= sites-heaps+initial
        // foldmax <= sites-initial
        fold[initial] = new cost[foldmax+1];
        cost acum = 0, barge = 0;
        fold[initial][0] = 0;
        for(unsigned int final = initial; final < initial+foldmax; final++) {
            // initial <= final < sites
            // final-initial+1 <= foldmax
            barge += weight[final];
            acum += barge * distance[final];
            fold[initial][final-initial+1] = acum;
        }
    }
}

inline cost Heaps::foldCost(unsigned int initial, unsigned int length) {
    assert(initial+length < sites);
    return fold[initial][length];
}


inline cost Heaps::solveSubproblem(unsigned int sites, unsigned int heaps) {
    assert(sites >= heaps);
    assert(sites <= this->sites);
    assert(heaps != 0 ^ sites == 0);    // sites == 0 iff heaps == 0 
    // Have to fold a total of (sites-heaps) times
    if(sites == heaps) {
        // Skip the remaining sites
        // This is a shortcut. The condition is also handled by 
        // successive a 0-length folds
        return 0;
    }
    // sites > 0
    // heaps > 0
    if(heaps == 1) {
        // 0 <= this->sites-sites < this->sites
        // 0 <= this->sites-sites+sites-1 = this->sites-1 < this->sites
        //return foldCost(this->sites-sites, sites-1);
        return fold[this->sites-sites][sites-1];
    }
    // heaps > 1
    cost minCost = COST_MAX;
    int newCost;
    //for(unsigned int folds=0; folds <= sites - heaps; folds++) {
    for(
        // 0 <= folds <= sites-heaps
        // 0 <= this->sites-sites < this->sites
        // 0 <= this->sites-sites+folds <= this->sites-heaps <= this->sites
        // 0 <= sites-folds-1 <= sites - 1 <= this->sites - 1
        // 0 <= sites - heaps - folds <= this->sites
        // heaps-1 > 0
        //minCost = std::min(minCost, foldCost(this->sites-sites, folds)
        newCost = fold[this->sites-sites][folds] + solved[sites-folds-1]
                                                         [sites-heaps-folds];
        minCost = newCost < minCost ? newCost : minCost;
    }
    return minCost;
}

cost Heaps::solve(unsigned int heaps) {
    calcFolds(heaps);
    temp = new cost[this->sites+1];
    solved = new cost*[this->sites + 1];
    cost *costs = new cost[1+sites*(sites+1)/2];
    cost *current = costs;
    solved[0] = current++;
    solved[0][0] = 0;
    for(unsigned int sites=1; sites <= this->sites; sites++) {
        unsigned int maxhheaps = std::min(sites, heaps);
        solved[sites] = current;
        current += sites;
        // 1 <= sites <= this->sites
        // Subproblem (n,k) makes reference to (n-f-1,k-1)
        // So (first arg-second arg) <= (this->sites - heaps)
        for(unsigned int hheaps=1; hheaps <= maxhheaps; hheaps++) {
            // 1 <= hheaps <= sites
            // 1 <= hheaps <= heaps
            // 0 <= sites-hheaps <= sites-1
            solved[sites][sites-hheaps] = solveSubproblem(sites, hheaps);
        }
    }
    cost ret = solved[this->sites][this->sites-heaps];
    delete solved;
    delete costs;
    delete temp;
    return ret;
}
/*
int main(int argc, char *argv[]) {

    while(true) {
        unsigned int sites, heaps;
        scanf("%d %d\n", &sites, &heaps);
        if(feof(stdin))
            break;
        cost *weight = new cost[sites];
        cost *distance = new cost[sites-1];
        int last_distance, this_distance;
        scanf("%d %d\n", &last_distance, &weight[0]);
        for(unsigned int i=1; i<sites; i++) {
            scanf("%d %d\n", &this_distance, &weight[i]);
            distance[i-1] = this_distance - last_distance;
            last_distance = this_distance;
        }
        printf("%d\n", Heaps(sites, weight, distance).solve(heaps));
        delete weight;
        delete distance;
    }
}
*/
